BeardedSpice/Tabs/QuodLibetTabAdapter.m

```diff
@@ -0,0 +1,382 @@
+#define FIFO_TIMEOUT            500 * NSEC_PER_MSEC // time until we give up on a response from QL, in nanoseconds
+
+// Control commands for QuodLibet
+#define QL_CONTROL_STATUS       @"status"
+#define QL_CONTROL_PLAY_PAUSE   @"play-pause"
+#define QL_CONTROL_PAUSE        @"pause"
+#define QL_CONTROL_NEXT         @"next"
+#define QL_CONTROL_PREVIOUS     @"previous"
+
+#define QL_CONTROL_JOINER       @"\0"
+NSString *const NSSTRING_EMPTY  = @"";
+
+#define QL_STATUS_PLAYING       @"playing"
+#define QL_STATUS_PAUSED        @"paused"
+
+// tag names in QL_CURRENT we are interested in
+#define QL_TRACK_TITLE          @"title"
+#define QL_TRACK_ARTIST         @"artist"
+#define QL_TRACK_ALBUM          @"album"
+#define QL_TRACK_FILENAME       @"~filename"
+#define QL_TRACK_LENGTH         @"~#length"
+
+
+
+static NSString * FormatNSTimeInterval(NSTimeInterval interval) {
+    NSInteger ti = (NSInteger)interval;
+    if (ti < 3600) {
+        return [NSString stringWithFormat:@"%02ld:%02ld", (long)(ti / 60), (long)(ti % 60)];
+    } else {
+        NSInteger min = ti / 60;
+        return [NSString stringWithFormat:@"%02ld:%02ld:%02ld", (long)(min / 60), (long)(min % 60), (long)(ti % 60)];
+    }
+}
+
+
+@implementation QuodLibetTabAdapter{
+    QuodLibetStatus _status;
+    double _progress;
+    BSTrack *_track;
+}
+
+- (id)init {
+    self = [super init];
+    if (self) {
+        _status = QuodLibetStatusUnknown;
+        _progress = 0;
+        _track = nil;
+    }
+    return self;
+}
+
++ (NSString *)displayName{
+    return APPNAME_QUOD_LIBET;
+}
+
++ (NSString *)bundleId{
+    return APPID_QUOD_LIBET;
+}
+
+- (NSString *)URL{
+    return APPNAME_QUOD_LIBET;
+}
+
+- (NSString *)key{
+    return @"A:" APPNAME_QUOD_LIBET;
+}
+
+- (NSString *)title{
+    @autoreleasepool {
+        [self loadQuodLibetState];
+
+        NSString *title;
+        if (_status == QuodLibetStatusPlaying) {
+
+            BSTrack *trackInfo = [self trackInfo];
+            if (trackInfo.track)
+                title = trackInfo.track;
+
+            if (trackInfo.artist) {
+
+                if (title) title = [title stringByAppendingFormat:@" - %@", trackInfo.artist];
+                else
+                    title = trackInfo.artist;
+            }
+        }
+
+        if ([NSString isNullOrEmpty:title]) {
+            title = NSLocalizedString(@"No Track", @"QuodLibetTabAdapter");
+        }
+
+        return [NSString stringWithFormat:@"%@ (%@)", title, APPNAME_QUOD_LIBET];
+    }
+}
+
+#pragma mark Player control methods
+
+- (void)toggle{
+    [self sendControlToQuodLibet:QL_CONTROL_PLAY_PAUSE];
+}
+
+- (void)pause{
+    [self sendControlToQuodLibet:QL_CONTROL_PAUSE];
+}
+
+- (void)next{
+    [self sendControlToQuodLibet:QL_CONTROL_NEXT];
+}
+
+- (void)previous{
+    [self sendControlToQuodLibet:QL_CONTROL_PREVIOUS];
+}
+
+- (BSTrack *)trackInfo{
+    [self loadQuodLibetState];
+    if (_status == QuodLibetStatusPlaying) {
+        if (_track) {
```

Comment:
@mjpieters This optimization will lead to the fact that the track data does not change even when playing another track, it seems to me. We need to receive up-to-date track information. :)